core/utils/cuckoo_map.h

```diff
@@ -0,0 +1,331 @@
+      Entry& entry = entries_[idx];
+
+      if (C(key, entry.first)) {
+        bucket.hash_values[i] = 0;
+        entry = Entry();
+        PushFreeKeyIndex(idx);
+        cnt_--;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  Entry* AddEntry(uint32_t primary, uint32_t secondary, const K& key,
+                  const V& value) {
+    uint32_t primary_bucket_index, secondary_bucket_index;
+    Entry* entry = nullptr;
+  again:
+    primary_bucket_index = primary & bucket_mask_;
+    if ((entry = AddToBucket(primary_bucket_index, key, value)) != nullptr) {
+      return entry;
+    }
+
+    secondary_bucket_index = secondary & bucket_mask_;
+    if ((entry = AddToBucket(secondary_bucket_index, key, value)) != nullptr) {
+      return entry;
+    }
+
+    if (MakeSpace(primary_bucket_index, 0) >= 0) {
+      goto again;
+    }
+
+    if (MakeSpace(secondary_bucket_index, 0) >= 0) {
+      goto again;
+    }
+
+    return nullptr;
+  }
+
+  int FindEmptySlot(uint32_t bucket_idx) {
+    const Bucket& bucket = buckets_[bucket_idx];
+    for (int i = 0; i < kEntriesPerBucket; i++) {
+      if (bucket.hash_values[i] == 0) {
+        return i;
+      }
+    }
+    return -1;
+  }
+
+  int MakeSpace(uint32_t index, int depth) {
+    if (depth >= kMaxCuckooPath) {
+      return -1;
+    }
+
+    Bucket& bucket = buckets_[index];
+
+    for (int i = 0; i < kEntriesPerBucket; i++) {
+      const K& key = bucket.key_indices[i];
+      uint32_t pri = HashNonZero(key);
+      uint32_t sec = HashSecondary(pri);
+
+      uint32_t alt_index;
+
+      // this entry is in its primary bucket?
+      if (pri == bucket.hash_values[i]) {
+        alt_index = sec & bucket_mask_;
+      } else if (sec == bucket.hash_values[i]) {
+        alt_index = pri & bucket_mask_;
+      } else {
+        return -1;
+      }
+
+      // Find empty slot
+      int j = FindEmptySlot(alt_index);
+      if (j == -1) {
+        j = MakeSpace(alt_index, depth + 1);
+      }
+      if (j >= 0) {
+        Bucket& alt_bucket = buckets_[alt_index];
+        alt_bucket.hash_values[j] = bucket.hash_values[i];
+        alt_bucket.key_indices[j] = bucket.key_indices[i];
+        bucket.hash_values[i] = 0;
+        return i;
+      }
+    }
+
+    return -1;
+  }
+
+  Entry* GetHash(uint32_t primary, const K& key) {
+    Entry* ret = GetFromBucket(primary, primary, key);
+    if (ret) {
+      return ret;
+    }
+    return GetFromBucket(primary, HashSecondary(primary), key);
+  }
+
+  Entry* GetFromBucket(uint32_t primary, uint32_t hash_val, const K& key) {
+    Bucket& bucket = buckets_[hash_val & bucket_mask_];
+    for (int i = 0; i < kEntriesPerBucket; i++) {
+      if (primary != bucket.hash_values[i]) {
+        continue;
+      }
+      uint32_t idx = bucket.key_indices[i];
+      if (C(entries_[idx].first, key)) {
+        return &entries_[idx];
+      }
+    }
+    return nullptr;
+  }
+
+  static uint32_t MakeNonZero(uint32_t v) {
```

Comment:
`HashNonZero` / `MakeNonZero()` are a legacy of `GetBulk()` function (which performs N lookups in parallel with SIMD instructions). SInce we don't support `GetBulk()` any longer (it didn't give solid performance boost), we can remove them and save a few cycles.